### PR TITLE
fix: select a single project to deploy

### PR DESCRIPTION
**Pull request by ptab**
90% of this PR is adding unit tests to the most important class of the project that was not covered in any meaningful way 😓 

Validated in:
* https://github.com/Vandebron/gh-mpyl/pull/101
* https://github.com/Vandebron/Vandebron/pull/17327

### CODE REVIEW

File: src/mpyl/run_plan.py [hunk 1]
```diff
@@ -47,11 +48,11 @@ def select_stage(self, stage: Stage) -> "RunPlan":
             selected_plan={stage: self.get_projects_for_stage(stage)},
         )
 
-    def select_projects(self, projects: set[Project]) -> "RunPlan":
+    def select_project(self, project: Project) -> "RunPlan":
         selected_plan = {}
 
         for stage, executions in self.selected_plan.items():
-            selected_plan[stage] = {e for e in executions if e.project in projects}
+            selected_plan[stage] = {e for e in executions if e.project == project}
```
[the remark this answers is not in the document]
Comment (ptab) [reply src/mpyl/run_plan.py:55]: Actually that's a very good point, that would even prevent us from having to load all projects again! 👍

[the remark this answers is not in the document]
Comment (ptab) [reply src/mpyl/run_plan.py:64]: Look [who's failing](https://github.com/Vandebron/gh-mpyl/actions/runs/12056093504/job/33618107042) now that we've implemented this check 😂

File: src/mpyl/run_plan.py [hunk 2]
```diff
@@ -25,49 +25,77 @@
...
+    def select_stage(self, stage_name: str) -> "RunPlan":
+        stage = None
+        for s in self._get_all_stages(use_full_plan=False):
+            if s.name == stage_name:
+                stage = s
+                break
+        if not stage:
+            raise ValueError(
+                f"Unable to select stage outside of the run plan: '{stage_name}'"
+            )
```
Comment (Jorg88) [inline src/mpyl/run_plan.py:52]:
```suggestion
        selected_stage = None
        for stage in self._get_all_stages(use_full_plan=False):
            if stage.name == stage_name:
                selected_stage = stage
                break
        if not selected_stage:
            raise ValueError(
                f"Unable to select stage outside of the run plan: '{stage_name}'"
            )
```

File: src/mpyl/run_plan.py [hunk 3]
```diff
@@ -25,49 +25,77 @@
...
+    def select_project(self, project_name: str) -> "RunPlan":
+        project = None
+        for p in self._get_all_projects(use_full_plan=False):
+            if p.name == project_name:
+                project = p
+                break
+        if not project:
+            raise ValueError(
+                f"Unable to select project outside of the run plan: '{project_name}'"
+            )
```
Comment (Jorg88) [inline src/mpyl/run_plan.py:69]:
```suggestion
        selected_project = None
        for project in self._get_all_projects(use_full_plan=False):
            if project.name == project_name:
                selected_project = project
                break
        if not selected_project:
            raise ValueError(
                f"Unable to select project outside of the run plan: '{project_name}'"
            )
```

File: src/mpyl/run_plan.py [hunk 4]
```diff
@@ -25,49 +25,77 @@
...
 
-        for stage, executions in self.selected_plan.items():
-            selected_plan[stage] = {e for e in executions if e.project in projects}
+        selected_plan = {}
+        for stage in self._get_all_stages(use_full_plan=False):
+            filtered = {
+                e
+                for e in self.get_projects_for_stage(stage, use_full_plan=False)
+                if e.name == project_name
```
Comment (Jorg88) [inline src/mpyl/run_plan.py:76]:
```suggestion
                project_execution
                for project_execution in self.get_projects_for_stage(stage, use_full_plan=False)
                if project_execution.name == project_name
```

File: src/mpyl/run_plan.py [hunk 5]
```diff
@@ -25,49 +25,77 @@
...
+            for project_execution in self._get_all_projects(use_full_plan=False)
         )
 
-    def get_all_projects(self, use_full_plan: bool = False) -> set[ProjectExecution]:
+    def _get_all_stages(self, use_full_plan: bool = False) -> set[Stage]:
+        if use_full_plan:
+            return set(self._full_plan.keys())
+        return set(self.selected_plan.keys())
+
+    def _get_all_projects(self, use_full_plan: bool = False) -> set[ProjectExecution]:
```
Comment (Jorg88) [inline src/mpyl/run_plan.py:98]: `_get_all_project_executions`?

File: tests/test_run_plan.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+            pytest.param(False, id="include_cached_projects=False"),
+        ],
+    )
+    def test_has_projects_to_run(self, include_cached_projects):
+        assert not self.run_plan.has_projects_to_run(
+            include_cached_projects=include_cached_projects,
+        )
+
+
+class TestFullRunPlan:
```
Comment (Jorg88) [inline tests/test_run_plan.py:110]: would it be worth considering putting each class in its own file?

**Review by Jorg88 — COMMENTED** [review]
Very nice job on all the extra test coverage!

File: src/mpyl/run_plan.py (hunk 5 above)
Comment (ptab) [reply src/mpyl/run_plan.py:98]: This method already existed, I just added the private `_` indicator behind it. I do like that name better though 👍

File: tests/test_run_plan.py (hunk 6 above)
Comment (ptab) [reply tests/test_run_plan.py:110]: Just to check: I thought the convention was to merely prefix test files with `test_`:
* `project.py` → `test_project.py`
* `run_plan.py` → `test_run_plan.py`
* `validation.py` → `test_validation.py`

And so on.

Comment (Jorg88) [reply tests/test_run_plan.py:110]: Ah fair point, I forgot about that!

File: actions/generate-kubernetes-manifests/action.yaml [hunk 7]
```diff
@@ -29,18 +29,18 @@ inputs:
     required: false
 runs:
   using: docker
-  image: docker://public.ecr.aws/vdb-public/gh-mpyl:v0.13.0
+  image: docker://public.ecr.aws/vdb-public/gh-mpyl:pr-101
```
Comment (Jorg88) [inline actions/generate-kubernetes-manifests/action.yaml:32]: Mandatory reminder to revert this before merging

**Review by Jorg88 — APPROVED** [review]

**Pull request merged**